core/client/src/main/java/alluxio/client/block/UnderStoreBlockInStream.java

```diff
@@ -145,6 +156,34 @@ public int read(byte[] b, int off, int len) throws IOException {
   }
 
   @Override
+  public int read(long pos, byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(PACKET_STREAMING_ENABLED,
+        "PositionedReadable interface is implemented only if packet streaming is enabled.");
+
+    if (pos >= mLength) {
+      return -1;
+    }
+
+    if (mUnderStoreStream instanceof PositionedReadable) {
+      return ((PositionedReadable) mUnderStoreStream).read(pos, b, off, len);
+    }
+    if (mUnderStoreStream instanceof org.apache.hadoop.fs.PositionedReadable) {
+      return ((org.apache.hadoop.fs.PositionedReadable) mUnderStoreStream).read(pos, b, off, len);
+    }
+
+    // This happens only when UFS delegation is off and the UFS is not HDFS.
+    synchronized (this) {
+      long oldPos = mPos;
+      try {
+        seek(pos);
+        return mUnderStoreStream.read(b, off, len);
```

Comment:
This section of code implements `PositionReadable` for all UFS input streams which do not have the direct implementation right? If so, should we move this logic to a wrapper (unless we only need to do this once)?